File: RxHttp/src/main/java/com/github/kongpf8848/rxhttp/ProgressRequestBody.java
package com.github.kongpf8848.rxhttp;

import com.github.kongpf8848.rxhttp.callback.ProgressCallback;

import java.io.IOException;

import okhttp3.MediaType;
import okhttp3.RequestBody;
import okio.Buffer;
import okio.BufferedSink;
import okio.ForwardingSink;
import okio.Okio;
import okio.Sink;


public class ProgressRequestBody extends RequestBody {
    private final RequestBody requestBody;
    private final ProgressCallback callback;
    private BufferedSink progressSink;


    public ProgressRequestBody(RequestBody requestBody, ProgressCallback callback) {
        this.requestBody = requestBody;
        this.callback = callback;
    }

    @Override
    public MediaType contentType() {
        return requestBody.contentType();
    }

    @Override
    public long contentLength() throws IOException {
        return requestBody.contentLength();
    }

    @Override
    public void writeTo(BufferedSink sink) throws IOException {
        if (callback == null) {
            requestBody.writeTo(sink);
            return;
        }
        if (progressSink == null) {
            progressSink = Okio.buffer(sink(sink));
        }
        callback.onProgress(0,contentLength());
        requestBody.writeTo(progressSink);
        progressSink.flush();
        progressSink.close();
    }

    private Sink sink(Sink sink) {
        return new ForwardingSink(sink) {
            long totalBytesWrite = 0L;
            long lastBytesWrite=0L;
            @Override
            public void write(Buffer source, long byteCount) throws IOException {
                super.write(source, byteCount);
                lastBytesWrite=totalBytesWrite;
                totalBytesWrite += byteCount;
                if(totalBytesWrite!=lastBytesWrite) {
                    if (callback != null) {
                        callback.onProgress(totalBytesWrite, contentLength());
                    }
                }
            }
        };
    }
}

File: RxHttp/src/main/java/com/github/kongpf8848/rxhttp/ProgressRequestBody.kt
package com.github.kongpf8848.rxhttp

import com.github.kongpf8848.rxhttp.callback.ProgressCallback
import okhttp3.MediaType
import okhttp3.RequestBody
import okio.*
import java.io.IOException

class ProgressRequestBody(private val requestBody: RequestBody, private val callback: ProgressCallback?) : RequestBody() {

    private var progressSink: BufferedSink? = null
    private var contentLength = 0L

    override fun contentType(): MediaType? {
        return requestBody.contentType()
    }

    @Throws(IOException::class)
    override fun contentLength(): Long {
        if (contentLength == 0L) {
            contentLength = requestBody.contentLength()
        }
        return contentLength
    }

    @Throws(IOException::class)
    override fun writeTo(sink: BufferedSink) {
        if (callback == null) {
            requestBody.writeTo(sink)
            return
        }
        if (progressSink == null) {
            progressSink = Okio.buffer(sink(sink))
        }
        callback.onProgress(0, contentLength())
        requestBody.writeTo(progressSink)
        progressSink?.apply {
            flush()
            close()
        }
    }

    private fun sink(sink: Sink): Sink {
        return object : ForwardingSink(sink) {
            var totalBytesWrite = 0L
            var lastBytesWrite = 0L

            @Throws(IOException::class)
            override fun write(source: Buffer, byteCount: Long) {
                super.write(source, byteCount)
                lastBytesWrite = totalBytesWrite
                totalBytesWrite += byteCount
                if (totalBytesWrite != lastBytesWrite) {
                    callback?.onProgress(totalBytesWrite, contentLength())
                }
            }
        }
    }

}